README.MD
# Bitcoin Live Price Wallpaper

## What is it?

This project is an HTML5 based web app that displays the current price of Bitcoin, updated every 5 seconds (configurable).
This is meant to be run as a wallpaper, using [Wallpaper Engine](https://www.wallpaperengine.io/), but can also be used standalone in the browser (the difference being that if you want to change the configurable properties, you'll have to manually edit them at the top of main.js since Wallpaper Engine provides a GUI with event hooks for property changes).

Anyone who has wallpaper engine and wants to use it can find it on the Steam Workshop [here](https://steamcommunity.com/sharedfiles/filedetails/?id=2573360835).

## How it works

This works by fetching the current Bitcoin-USD price pair from CoinGecko every {refreshInterval} seconds. It then changes the displayed price to the newly fetched one, using an animation when doing so as this is a wallpaper.

## Roadmap

Note the wallpaper is currently very minimal. I want to maintain the minimal look, but add more features that can be toggled on via properties.
My first plan is to add support for other fiat currencies, likely via a configurable property specifying the currency ISO code. I will publish a more detailed roadmap soon.

## Contributing

This is a standard HTML5 stack project, so anyone who knows HTML, CSS, and JavaScript should be able to make contributions.
I plan to make a more detailed CONTRIBUTING.md in the future, but for the now the following guidelines apply:

* Problems should be reported via GitHub issues
* Features and changes may be discussed via GitHub issues
* Small features that can be toggled on and off via properties with no effects on other features can be submitted via pull request.
* Larger features or major changes to existing features should be discussed via a GitHub issue first.
* Quality of life improvements to existing code and/or documentation is very welcome and can be submitted via pull request.
* Please make your pull request against the development branch. This branch gets merged into master when doing a release for Wallpaper Engine (Steam Workshop).

All of the project code is documented and should be self-explanatory (if it's not feel free to open an issue). For Wallpaper Engine related things, see their [documentation](https://docs.wallpaperengine.io/en/web/overview.html).

If you like to test this project in Wallpaper Engine, you can place it in C:\Program Files (x86)\Steam\steamapps\common\wallpaper_engine\projects\myprojects or the equivalent on your system, open Wallpaper Engine, go to the bottom, click open wallpaper, and then open offline wallpaper (animated), and select the project.json file that's in this repo. This is only necessary when modifying wallpaper-specific code (such as configurable properties), otherwise debugging in the web browser is sufficient.

Releases onto Steam Workshop for Wallpaper Engine users is done manually by Andrew Mitchell, as this is hard to automate and releases will likely roll up multiple new commits rather than a new release per commit approach.

## License

This project and any assets contributed to it, are licensed under the MIT license. See LICENSE.md for details.


CHANGELOG.MD
# Changelog

## 9/2/2021

- Clear counter before starting new one
- Modified count animation in attempt to reduce bug where count drops below zero randomly
- Better code documentation
- Open sourced the project! See [https://github.com/realandrew/btc-live-wallpaper](https://github.com/realandrew/btc-live-wallpaper).

## 8/19/2021.3

- Fixed a bug resulting in duplicated network requests. Network traffic should now be halved (and this possibly fixes a bug).

## 8/19/2021.2

- Implemented toastify.js for non-distracting debug/error notifications, rather than using alert() which is distracting and is usually hidden behind all your windows.

## 8/19/2021.1

- Fixed bug where background wasn't changing colors.
- Disabled refresh bar by default
- Added some debug alerts to try to narrow down bug where price drops to negative infinity.

## 8/13/2021.5

- Fixed the refresh bar enabled property not working.
- Switching one property now stops reloading of other properties (i.e. now the background doesn't switch when flipping the refresh enabled property).

## 8/13/2021.4

- Refresh bar to show when next price refresh is (a little bit glitchy on higher refresh intervals, this is being looked into).
- Fixed bug where wallpaper engine properties sometimes weren't applied.

## 8/13/2021.3

- Switched from deprecated CoinDesk API to the CoinGecko API (removed decimal places as well since CoinGecko returns an integer).
    This is in preparation for future features, which require more data that CoinDesk's API didn't return.
- Added a simple dark mode option (might be improved/expanded in the future).

## 8/13/2021.2

- Added text shadow.
- Added a wallpaper engine property for price refresh interval.

## 8/13/2021

- Fixed price being interpreted as an integer, causing decimal to always end in .00.

## 8/12/2021.3

- Added background transitions. The background will transition between gradiants every 60 seconds.
    Currently there are only 4 gradients, but more will be added in the future.
- Possibly fixed bug where upon resumption of playing wallpaper the counter would go crazy.

## 8/12/2021.2

- Updated font.
- Adjusted footer css to make sure CoinDesk link is white whether visited or not.

## 8/12/2021

- Initial release.


js/countup.js
/* Count-up animation adapted from https://jshakespeare.com/simple-count-up-number-animation-javascript-react/ */

const animationDuration = 2000; // How long you want the animation to take, in ms
const frameDuration = 1000 / 60; // Calculate how long each ‘frame’ should last if we want to update the animation 60 times per second
const totalFrames = Math.round(animationDuration / frameDuration); // Use that to calculate how many frames we need to complete the animation
const easeOutQuad = t => t * (2 - t); // An ease-out function that slows the count as it progresses

let counter; // Interval instance
let frame = 0; // Current animation frame
let countFrom; // Number to start count animation from
let countTo; // Number to end count animation at

/**
 * Animates a number element
 * @param {*} el - The element containing the numbers to animate the change of.
 * @param {number} oldNumber The current value of the number.
 * @param {number} newNumber The value to change the number to.
 * @param {number} decimalPlaces The amount of decimal places to display.
 * @returns 
 */
const animateCount = (el, oldNumber, newNumber, decimalPlaces = 0) => {
	if (oldNumber == newNumber) { return };
	countFrom = parseFloat(oldNumber, 10);
	countTo = parseFloat(newNumber, 10);
	if (frame == 0) {
		let currentCount = countFrom;
		// Start the animation running 60 times per second
		clearInterval(counter);
		counter = setInterval(() => {
			frame++;
			// Calculate our progress as a value between 0 and 1
			// Pass that value to our easing function to get our
			// progress on a curve
			const progress = easeOutQuad(frame / totalFrames);
			// Use the progress value to calculate the current count
			currentCount = Number(countFrom + ((countTo - countFrom) * progress)).toFixed(decimalPlaces).replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');
	
			// If the current count has changed, update the element
			if (parseInt(el.innerHTML, 10) !== currentCount) {
				el.innerHTML = currentCount;
			}
	
			// If we’ve reached our last frame, stop the animation
			if (frame >= totalFrames) {
				clearInterval(counter);
				frame = 0;
			}
		}, frameDuration);
	}
};

js/backgroundAnim.js
/* Copyright 2021 Andrew Mitchell */

/* Settings */
const switchTime = 60000; // Time between transitions in milliseconds

// Instance variables
let opacity = 0; // Since CSS doesn't support transitions, we actually just have 2 gradients that we transition the opacity of.
let switchInterval; // Interval instance used for switching between gradients.

// Array of gradients we can switch the background to
const gradients = [
    { gradient: "linear-gradient(90deg, hsla(333, 100%, 53%, 1) 0%, hsla(33, 94%, 57%, 1) 100%)", fallback: "hsla(333, 100%, 53%, 1)" },
    { gradient: "linear-gradient(90deg, hsla(217, 100%, 50%, 1) 0%, hsla(186, 100%, 69%, 1) 100%)", fallback: "hsla(217, 100%, 50%, 1)" },
    { gradient: "linear-gradient(90deg, hsla(33, 100%, 53%, 1) 0%, hsla(58, 100%, 68%, 1) 100%)", fallback: "hsla(33, 100%, 53%, 1)"},
    { gradient: "linear-gradient(90deg, hsla(152, 100%, 50%, 1) 0%, hsla(186, 100%, 69%, 1) 100%)", fallback: "hsla(152, 100%, 50%, 1)"}
];

// A custom event we call when a wallpaper engine property is changed.
window.addEventListener('weLoaded', (evt) => {
    if ((evt.detail.darkMode != evt.detail.previousProperties.darkMode) || !evt.detail.hasLoaded) // Only run if the darkmode value had changed
    {
        if (evt.detail.darkMode) { // Toggle on dark mode
            clearInterval(switchInterval);
            document.body.style.background = "#000";
            document.body.style.setProperty('--background-fallback', "#000");
            document.body.style.setProperty('--background-gradient', "#000");
            document.body.style.setProperty('--before-fallback', "#000");
            document.body.style.setProperty('--before-gradient', "#000");
        } else { // Toggle off dark mode
            clearInterval(switchInterval);
            SwitchToNewGradient();
            switchInterval = setInterval(() => {
                SwitchToNewGradient();
            }, switchTime);
        }
    }
});

/**
 * Transitions the background to a new gradient
 */
function SwitchToNewGradient() {
    // Pick a new gradient to switch to
    let index = RandomArrayIndex(gradients.length);
    let gradientObj = gradients[index];

    // We have 2 gradients we switch between. Since CSS doesn't support gradient transitions we just transition the opacity for a similar effect.
    opacity = opacity == 1 ? 0 : 1;
    if (opacity) { // Update the first gradient
        document.body.style.setProperty('--before-fallback', gradientObj.fallback);
        document.body.style.setProperty('--before-gradient', gradientObj.gradient);
    } else { // Update the other gradient
        document.body.style.setProperty('--background-fallback', gradientObj.fallback);
        document.body.style.setProperty('--background-gradient', gradientObj.gradient);
    }
    document.body.style.setProperty('--before-opacity', opacity); // Toggle opacity, causing the transition.
}

/**
 * Returns a random index for an array of a specified length.
 * @param {Number} length Length of the array.
 * @returns {Number} Random index.
 */
// Math.random give random between [0 - 1), exclusive of 1. This is multiplied by length to get a random between (0 - length). Then floored to give (0 - (length - 1)).
function RandomArrayIndex(length) {
    return Math.floor(Math.random() * length);
}

js/main.js
/* Copyright 2021 Andrew Mitchell */

// Configurable properties
let refreshInterval = 5; // How often to refresh (in seconds). Set from wallpaper engine settings
let darkMode = false; // Whether to use a black background instead of the animated color gradients. Set from wallpaper engine settings.

// Instance globals
let previousProperties = { refreshInterval: refreshInterval, darkMode: darkMode }; // Stores previous settings so that changes can be detected.
let hasLoaded = false; // Whether wallpaper engine property hooks have been loaded.
let lastPrice = 0.0; // The last price of Bitcoin
let refreshTimeout; // Stores the timeout instead until the next refresh.
let _btcusd; // The html element containing the USD price of Bitcoin.

// Check if running within wallpaper engine, if not then set hasLoaded to true
window.wallpaperPropertyListener = {
    applyUserProperties: function(properties) {
        if (properties.refreshinterval) { // Get the refreshInterval setting and apply it
            refreshInterval = properties.refreshinterval.value;
            StartBTCInterval(i = 0, ((refreshInterval * 1000) / 100));
        }
        if (properties.darkmode) { // Get the dark mode setting and apply it
            darkMode = properties.darkmode.value;
        }
        // A custom event is dispatched to let our our scripts know a wallpaper engine property was changed.
        const evt = new CustomEvent('weLoaded', { detail: { hasLoaded: hasLoaded, refreshInterval: refreshInterval, darkMode: darkMode, previousProperties: previousProperties }});
        window.dispatchEvent(evt);
        previousProperties = { refreshInterval: refreshInterval, darkMode: darkMode }; // Update the previous properties to the new properties.
        hasLoaded = true; // We have now loaded
    },
};

// Start the counter once the window has loaded
window.addEventListener('load', async (event) => {
    _btcusd = document.getElementById("usd"); // Set the counter element
    RefreshBTC(); // This runs the price check of bitcoin once to get us our initial price
    StartBTCInterval(i = 0, ((refreshInterval * 1000) / 100)); // And here we start the interval that checks every x seconds based on the refreshInterval property.
});

// 
/**
 * This runs the price check interval, we actually use a timeout so that we can dynamically update if the refreshInterval property changes.
 * Works by recursively calling itself using a timeout of ()
 * @param {number} index Current index of the timeout loop
 * @param {number} iterations Number of iterations of the timeout loop
 * @returns The timeout interval instance, so it can be cleared before starting a new interval.
 */
async function StartBTCInterval(index, iterations) {
    clearTimeout(refreshInterval); // Stop the current timeout if it exists
    // Start a new timeout loop
    refreshInterval = setTimeout(function () {
        index++;
        if (index >= iterations) {
            RefreshBTC();
            index = 0;
        }
        refreshTimeout = StartBTCInterval(index, ((refreshInterval * 1000) / 100));
    }, 100);
    return refreshInterval;
}

/**
 * Refreshs the current BTC price.
 */
async function RefreshBTC() {
    try {
        let priceUSD = await GetCurrentBTCPrice("usd"); // Fetchs the current BTC price in USD
        DisplayBTCPrice(priceUSD); // Display the updated price
    } catch (err) {
        // Display a toast message notifying user of the error
        Toastify({
            text: "[Error] Failed to fetch price! " + err.message,
            duration: 5000,
            gravity: "bottom", // `top` or `bottom`
            position: "right", // `left`, `center` or `right`
            offset: {
                x: 30,
                y: 50
            },
        }).showToast();
    }
}

/**
 * Updates the displayed BTC price to the once specified.
 * @param {number} btcUSD 
 */
function DisplayBTCPrice(btcUSD) {
    animateCount(_btcusd, lastPrice, btcUSD); // Runs the animation effect to update to the new price.
    lastPrice = btcUSD; // Update the last price to be the new one, used for the animated effect starting count.
}

/**
 * Fetchs the current BTC price in the specified currency.
 * @param {string} currency The currency to used when fetching the price of a BTC-Currency pair. (ISO 4217 I believe, if not check the CoinGecko documentation).
 * @returns {number} BTC-Currency price.
 */
async function GetCurrentBTCPrice(currency) {
    return new Promise((resolve, reject) => {
        if (String(currency).toLowerCase() == "usd") {
            fetch('https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=bitcoin&order=market_cap_desc&per_page=100&page=1&sparkline=false&price_change_percentage=1h%2C24h%2C7d')
            .then((resp) => {
                if (resp.status != 200) {
                    reject(new Error("Error fetching price! Status: " + resp.status));
                } else {
                    return resp;
                }
            })
            .then(resp => resp.json())
            .then(data => {
                resolve(data[0].current_price);
            }).catch(err => {
                reject(err);
            });
        } else {
            reject(new Error("Invalid value for currency parameter!"));
        }
    });
}

/* Utility functions */
/**
 * Allows us to "sleep" by returning after the specified milliseconds.
 * @param {number} ms Milliseconds to sleep for.
 * @returns {Promise<void>}
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}